man/ply.1.ronn
ply(1) -- dynamically instrument the kernel
===========================================

## SYNOPSIS

`ply` <program-file> <br>
`ply` <program-text>

## DESCRIPTION

ply dynamically instruments the running kernel to aggregate and
extract user-defined data. It compiles an input program to one or more
Linux bpf(2) binaries and attaches them to arbitrary points in the kernel
using kprobes and tracepoints.

## OPTIONS

  * `-c` <command>, `--command`=<command>:
    When all probes are running, run <command>. When the command
    exits, stop all probes and exit. The command is run as if invoked
    with `sh -c <command>`.

  * `-d`, `--debug`:
    Enable debugging output.

  * `-e`, `--dry-run`:
    Exit after compilation, without actually instrumenting the
    system. Typically used in conjunction with `--dump`.

  * `-h`, `--help`:
    Print usage message.

  * `-k`, `--keep-going`:
    Instead of printing a warning and exiting whenever any trace data
    is lost, only print the warning and keep going.

  * `-S`, `--dump`:
    After compilation, dump the internal AST, generated BPF
    instructions and other internal information. This is very useful
    to include when reporting a bug.

  * `-v`, `--version`:
    Print version information.


## SYNTAX

The syntax is C-like in general, taking its inspiration dtrace(1) and,
by extension, from awk(1).


### Probes

A program consists of one or more _probes_, which are analogous to
awk's pattern-action statements. The syntax for a probe is as follows:

    provider:probe-definition ['/' predicate '/']
    {
         statement ';'
        [statement ';' ... ]
    }

The _provider_ selects which probe interface to use. See the PROVIDERS
section for more information about each provider. It is then up to the
provider to parse the _probe-definition_ to determine the point(s) of
instrumentation.

When tracing, it is often desirable to filter events to match some
criteria. Because of this, ply allows you to provide a _predicate_,
i.e. an expression that must evaluate to a non-zero value in order for
the probe to be executed.

Then follows a block of _statements_ that perform the actual information
gathering.

A provider may define a default probe clause to be used if the user
does not supply one.


### Control of Flow

Probes support basic conditional control of flow via an
_if-statement_, which conforms to the same rules as C's equivalent:

    'if' '(' expr ')'
        statement ';' | block
    [else
        statement ';' | block]

In order to ensure that a probe will have a finite run-time the kernel
does not allow backwards branching. As a result, ply does not have any
loop construct like _for_ or _while_. A simple _for_ statement with an
invariant that is known at compile-time could be added later. In that
case we could unroll the loop when generating BPF.


### Type System

The type system is modeled after C. As such ply understands the
difference between signed and unsigned integers, the difference
between a short and a long long, what separates an integer from a
pointer, how a struct is laid out in memory and so on. It is not
complete though, notably floating point numbers and unions are
missing.

Programs are statically typed, but all types are inferred
automatically. Thus, the type system is mostly hidden from the
user. Plans are to expose more of it in the future by allowing casts,
type declarations and so on.

Numbers and string literals are specified in the same way as in
C.


### Maps

The primary way to extract information is to store it in a _map_,
i.e. in a hash table. Like awk(1), ply dynamically creates any
referenced maps and their key and value types are inferred from the
context in which they are used. All maps are in the global scope and
can thus be used both for extracting data to the end-user, and for
carrying data between probes. Map names follow the rules of
identifiers from C.

    mapname[exprs]

Data can be stored in a map by assigning a value to a given key:

    mapname[exprs] = expr

The _delete_ keyword can be used to remove an association from a map:

    delete mapname[exprs]

You can also remove all elements in the map using _clear_ function.


### Aggregations

More often than not, looking at each individual datum from a trace is
not nearly as helpful as an _aggregation_ of the data. Therefore ply
supports aggregating data at the source, thereby reducing tracing
overhead. Aggregations are syntactically similar to maps, indeed they
are a kind of map, but they are distinguished by a leading '@'. Also,
they can only be assigned the result of one of the following
aggregation functions:

  * `@agg[exprs] = count()`:
    Bump a counter.

  * `@agg[exprs] = sum(scalr-expr)`:
    Evaluates the argument and aggregates the result.

  * `@agg[exprs] = quantize(scalar-expr)`:
    Evaluates the argument and aggregates on the most significant bit
    of the result. In other words, it stores the distribution of the
    expression.


## PROVIDERS

A _provider_ makes data available to the user by exporting functions
and variables to the probe. Function calls use the same syntax as most
languages that inherit from C. In addition to the provider-specific
functions, all providers inherits a set of common functions and
variables:

  * `char[16] comm`, `char[16] execname`
    _name_ of the running process's executable.

  * `u32 cpu`
    _CPU ID_ of the processor on which the probe fired.

  * `u32 gid`
    _Group ID_ of the running process.

  * `u32 kpid`:
    _Kernel PID_ of the running process. Also known as _pid_ by the
    kernel. For a single-threaded process _kpid_ is equal to
    _pid_. For multi-threaded processes, _kpid_ will be unique while
    _pid_ will be the same across all threads.

  * `char[N] mem(void *address [, int size])`
    Copy _size_ bytes from _address_. If _size_ is omitted, 64 bytes
    will be copied.

  * `s64 time`, `s64 walltime`:
    Nanoseconds elaped since system boot. _time_ is intended for time
    deltas and _walltime_ should be used for timestamps. They refer to
    the same data, but with different default output formats.

  * `u32 pid`:
    _Process ID_ of the running process. Also known as _thread group
    ID_ (tgid) by the kernel.

  * `void print(...)`:
    _Print_ each expression with its default output format, separated
    by commas and terminated with a newline, to ply's standard out.

  * `void printf(format, ...)`:
    Prints _formatted output_ to ply's standard out. In addition to
    the formats recognized by the printf sitting in your <stdio.h>,
    ply's also recognizes '%v' which will dump the value according to
    the inferred type's default (i.e. how _print_ would print it).

  * `int strcmp(char *a, char *b)`:
    Returns -1, 0 or 1 if the first argument is less than, equal to or
    greater than the second argument respectively. Strings are
    compared by their lexicographical order.

  * `u32 uid`:
    _User ID_ of the running process.


### kprobe and kretprobe

These providers use the corresponding kernel features to instrument
arbitrary instructions in the kernel. The _probe-definition_ may be
either an address or a symbol name. When using a symbol name, glob
expansion is performed allowing a single probe to be inserted at
multiple locations. An offset relative to a symbol may also be
specfied for kprobes.

Examples:

  * _kretprobe:schedule_: Trace every time `schedule` returns.
  * _kprobe:SyS\_*_: Trace every time a syscall is made.
  * _kprobe:dev_hard_start_xmit+8_: Trace function with offset.


Shared variables:

  * `struct pt_regs *regs`:
    Hardware register contents from when the probe was triggered. This
    matches the definition in <sys/ptrace.h> on your system.

  * `u32 stack`:
    _Stack trace ID_ of the current probe. This is just returns an
    index into a separate map containing the actual instruction
    pointers. As a user though, you can think of this function as
    returning a string containing the stack trace at the current
    location. Indeed _print(stack)_ will produce exactly that.

    CAUTION: On some architectures (looking at you, ARM), capturing
    stack traces at the entry of a function, before the prologue has
    run, does not work. Setting your probe after the prologue will
    work around the issue (typically two instructions, or +8, on ARM).

_kprobe_ specific functions:

  * `arg0`, `arg1` ... `argN`:

    Returns the value of the specified _argument_ of the function to
    which the probe was attached, zero-indexed. I.e. arg0 is the 1st
    argument, arg1 is the 2nd, and so on.

    CAUTION: ply simply maps registers to arguments according to the
    syscall ABI. If your compiler decides to optimize out arguments or
    do other sneaky things, ply will be utterly oblivious to that.

  * `void *caller`:
    The program counter, as recorded in `regs`, at the time the probe
    was triggered.  was attached. The default output format will
    resolve it to a symbolic name if one is available.

_kretprobe_ specific function:

  * `retval`:
    Return value of the probed function.


### tracepoint

The tracepoint provider can instrument all stable tracepoints in the
kernel. They are identified by their relative path from the
`/sys/kernel/debug/tracing/events` directory, where each leaf
directory corresponds to a tracepoint.

Examples:

  * _tracepoint:sched/sched_wakeup_: Trace every time a process is
    awoken.
  * _tracepoint:irq/irq_handler_entry_: Trace every time an interrupt
    is handled.

Variables:

  * `struct <X> *data`:
    A struct is dynamically generated for each tracepoint by parsing
    its `format` file. I.e. if the contents of the format file looks
    like the following:

    ```
        name: tcp_send_reset
        ID: 1304
        format:
        	field:unsigned short common_type;	offset:0;	size:2;	signed:0;
        	field:unsigned char common_flags;	offset:2;	size:1;	signed:0;
        	field:unsigned char common_preempt_count;	offset:3;	size:1;	signed:0;
        	field:int common_pid;	offset:4;	size:4;	signed:1;

        	field:const void * skbaddr;	offset:8;	size:8;	signed:0;
        	field:const void * skaddr;	offset:16;	size:8;	signed:0;
        	field:__u16 sport;	offset:24;	size:2;	signed:0;
        	field:__u16 dport;	offset:26;	size:2;	signed:0;
        	field:__u8 saddr[4];	offset:28;	size:4;	signed:0;
        	field:__u8 daddr[4];	offset:32;	size:4;	signed:0;
        	field:__u8 saddr_v6[16];	offset:36;	size:16;	signed:0;
        	field:__u8 daddr_v6[16];	offset:52;	size:16;	signed:0;
    ```

    Then `data` would point to a struct of the following type:

    ```
        struct data {
            unsigned short common_type;
            unsigned char common_flags;
            unsigned char common_preempt_count;
            int common_pid;

            const void * skbaddr;
            const void * skaddr;
            __u16 sport;
            __u16 dport;
            __u8 saddr[4];
            __u8 daddr[4];
            __u8 saddr_v6[16];
            __u8 daddr_v6[16];
        };
    ```

Functions:

  * `char[N] dyn(void *address [, int size])`:
    Copy _size_ bytes from a dynamic data pointer in `data`, i.e. a
    member marked with `__data_loc`. If _size_ is omitted, the default
    string size determines the number of bytes to be copied.


### BEGIN and END

These special providers are called at the beginning and the end of the
tracing session like awk and bpftrace.  The names are case sensitive.
Users can print some messages or fill maps to known info.

### interval

The interval provider will be trigger at each given interval.
Users can specify time and unit (optional). If unit is omitted, then
second is used.  The supported units are:

  * m: minutes
  * s: seconds  (default)
  * ms: milli-seconds
  * us: micro-seconds
  * ns: nano-seconds

Examples:

  * _interval:1_: Called for every second
  * _interval:500ms_: Called for every 500 milli-second

### profile

The profile provider supports profiling by allowing the user to specify
how many times it will fire per-second.  Values of 1-1000 are supported,
and the profile provider supports two probe formats:

  * profile:[N]hz: Profile on all CPUs N times per second
  * profile:[C]:[N]hz: Profile on CPU C N times per second


## EXAMPLE

### Extracting data

Print all `openat`ed files on the system, and who opened them:

    kprobe:SyS_openat
    {
        print(comm, pid, str(arg1));
    }

### Quantize

Record the distribution of the return value of read(2):

    kretprobe:SyS_read
    {
        @["dist"] = quantize(retval);
    }


### Wildcards

Count all syscalls made on the system, grouped by function:

    kprobe:SyS_*
    {
        @[caller] = count();
    }


Count all syscalls made by every dd(1) process, grouped by function:

    kprobe:SyS_* / !strcmp(execname, "dd") /
    {
        @[caller] = count();
    }


### Object Tracking

Record the distribution of the time it takes an _skb_ to go from
_netif_receive_ to _ip_rcv_:

    kprobe:__netif_receive_skb_core
    {
        rx[arg0] = time;
    }

    kprobe:ip_rcv / rx[arg0] /
    {
        @["diff"] = quantize(time - rx[arg0]);
    }


## RETURN VALUE

  * `0`:
    Program was successfully compiled and loaded into the kernel.

  * `Non-Zero`:
    An error occurred during compilation or during kernel setup.


## AUTHORS

Tobias Waldekranz <tobias@waldekranz.com>

## COPYRIGHT

Copyright 2018 Tobias Waldekranz

License: GPLv2


## SEE ALSO

awk(1) dtrace(1) bpf(2)


src/ply/ply.c
/*
 * Copyright Tobias Waldekranz <tobias@waldekranz.com>
 *
 * SPDX-License-Identifier: GPL-2.0
 */
#include <assert.h>
#include <getopt.h>
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <linux/version.h>
#include <sys/resource.h>
#include <sys/wait.h>

#include <ply/ply.h>

/* Embedded self-test.sh */
extern char _binary_self_test_sh_start;
extern char _binary_self_test_sh_end;

static void usage()
{
	fputs("ply - Dynamic tracing utility\n"
	      "\n"
	      "Usage:\n"
	      "  ply [options] <ply-text>\n"
	      "  ply [options] <ply-file>\n"
	      "\n"
	      "Options:\n"
	      "  -c COMMAND     Run COMMAND in a shell, exit upon completion.\n"
	      "  -d             Enable debug output.\n"
	      "  -e             Exit after compiling.\n"
	      "  -h             Print usage message and exit.\n"
	      "  -k             Keep going in face of trace buffer overruns.\n"
	      "  -S             Show generated BPF.\n"
	      "  -T             Run self-test.\n"
	      "  -v             Print version information.\n",
	      stderr);
}

static void self_test(char *plybin)
{
	size_t self_test_sz;
	char *cmd;
	FILE *sh;

	self_test_sz = &_binary_self_test_sh_end - &_binary_self_test_sh_start;

	asprintf(&cmd, "PLYBIN=%s /bin/sh", plybin);
	sh = popen(cmd, "w");
	free(cmd);
	if (!sh)
		goto err;

	if (fwrite(&_binary_self_test_sh_start, self_test_sz, 1, sh) != 1)
		goto err;

	exit(pclose(sh) ? 1 : 0);

err:
	_e("unable to run self-test\n");
	exit(1);
}

static void memlock_uncap(void)
{
	struct rlimit limit;
	rlim_t current;
	int err;

	err = getrlimit(RLIMIT_MEMLOCK, &limit);
	if (err) {
		_e("unable to retrieve memlock limit, "
		   "maps are likely limited in size\n");
		return;
	}

	current = limit.rlim_cur;

	/* The total size of all maps that ply is allowed to create is
	 * limited by the amount of memory that can be locked into
	 * RAM. By default, this limit can be quite low (64kB on a
	 * standard x86_64 box running a recent kernel). So this
	 * simply tells the kernel to allow ply to use as much as it
	 * needs. */
	limit.rlim_cur = limit.rlim_max = RLIM_INFINITY;
	err = setrlimit(RLIMIT_MEMLOCK, &limit);
	if (err) {
		const char *suffix = "B";

		if (!(current & 0xfffff)) {
			suffix = "MB";
			current >>= 20;
		} else if (!(current & 0x3ff)) {
			suffix = "kB";
			current >>= 10;
		}

		_w("could not remove memlock size restriction\n");
		_w("total map size is limited to %lu%s\n", current, suffix);
		return;
	}

	_d("unlimited memlock\n");
}

void dump(struct ply *ply)
{
	struct ply_probe *pb;

	if (!ply->probes) {
		printf("NO PROBES\n");
		return;
	}

	printf("\n\n-- globals\n");
	symtab_dump(&ply->globals, stdout);

	ply_probe_foreach(ply, pb) {
		printf("%s\n", pb->probe ? : "<null>");

		if (pb->ast)
			ast_fprint(stdout, pb->ast);
		else
			printf("NO AST\n");

		printf("\n-- locals\n");
		symtab_dump(&pb->locals, stdout);
		printf("-- ir\n");
		ir_dump(pb->ir, stdout);
	}
}

static void version()
{
	printf("%s (linux-version:%u~%u.%u.%u)\n",
	       PACKAGE_STRING, LINUX_VERSION_CODE,
	       (LINUX_VERSION_CODE >> 16) & 0xff,
	       (LINUX_VERSION_CODE >>  8) & 0xff,
	       (LINUX_VERSION_CODE >>  0) & 0xff);
}

static const char *sopts = "c:dehkSTv";
static struct option lopts[] = {
	{ "command",    required_argument, 0, 'c' },
	{ "debug",      no_argument,       0, 'd' },
	{ "dry-run",    no_argument,       0, 'e' },
	{ "help",       no_argument,       0, 'h' },
	{ "keep-going", no_argument,       0, 'k' },
	{ "dump",       no_argument,       0, 'S' },
	{ "self-test",  no_argument,       0, 'T' },
	{ "version",    no_argument,       0, 'v' },

	{ NULL }
};

FILE *get_src(int argc, char **argv)
{
	if (!argc)
		return NULL;

	/* if the argument names an existing file that we have access
	 * to, use it as the source. */
	if (!access(argv[0], R_OK))
		return fopen(argv[0], "r");

	/* TODO concat multiple argvs to one string and parse that as
	 * a ply script */

	/* otherwise, parse the argument as a ply script. */
	return fmemopen(argv[0], strlen(argv[0]), "r");
}

int inferior_prep(const char *cmd, int *infpid, int *inftrig)
{
	int err, pid, trig[2];

	err = pipe(trig);
	if (err)
		return err;

	*inftrig = trig[1];

	pid = fork();
	if (pid < 0)
		return pid;

	if (pid) {
		char str[16];

		*infpid = pid;

		/* allow scripts to reference the pid of the inferior
		 * as $target. */
		snprintf(str, sizeof(str), "%d", pid);
		setenv("target", str, 0);
		return 0;
	}

	/* wait for parent to compile and get ready */
	if (read(trig[0], &err, sizeof(err)) != sizeof(err))
		return -EINVAL;

	/* if parent sends us an error, don't run the command. most
	 * probably the script did not compile. */
	if (err)
		exit(0);

	return execl("/bin/sh", "sh", "-c", cmd, NULL);
}

static int term_sig = 0;
static void term(int sig)
{
	term_sig = sig;
	return;
}
static const struct sigaction term_action = {
	.sa_handler = term,
	.sa_flags = 0,
};

void __attribute__((noinline)) ply_begin_trigger(void) { asm volatile (""); }
void __attribute__((noinline)) ply_end_trigger(void) { asm volatile (""); }

int main(int argc, char **argv)
{
	struct ply *ply;
	struct ply_return ret = { .err = 1 };
	int opt, infpid, inftrig;
	int f_dryrun, f_dump;
	FILE *src;
	char *cmd = NULL;

	f_dryrun = f_dump = 0;
	while ((opt = getopt_long(argc, argv, sopts, lopts, NULL)) > 0) {
		switch (opt) {
		case 'c':
			cmd = optarg;
			break;
		case 'd':
			ply_config.verify = 1;
			ply_debug = 1;
			break;
		case 'e':
			f_dryrun = 1;
			ply_config.ksyms = 0;
			break;
		case 'h':
			usage(); exit(0);
			break;
		case 'k':
			ply_config.strict = 0;
			break;
		case 'S':
			f_dump = 1;
			break;
		case 'T':
			self_test(argv[0]); exit(1);
			break;
		case 'v':
			version(); exit(0);
			break;

		default:
			_e("unknown option '%c'\n", opt);
			usage(); exit(1);
			break;
		}
	}

	src = get_src(argc - optind, &argv[optind]);
	if (!src) {
		_e("no input\n");
		usage(); exit(1);
	}

	if (cmd && inferior_prep(cmd, &infpid, &inftrig))
		exit(1);

	/* TODO figure this out dynamically. terminfo? */
	ply_config.unicode = 1;

	ply_init(ply_begin_trigger, ply_end_trigger);

	ply_alloc(&ply);
	ret.val = ply_fparse(ply, src);
	if (ret.val)
		goto err;

	ret.val = ply_compile(ply);

	if (f_dump)
		dump(ply);

	if (ret.val)
		goto err;

	if (f_dryrun)
		goto unload;

	memlock_uncap();

	ret = ply_load(ply);
	if (ret.exit || ret.err)
		goto err;

	ply_start(ply);
	_d("ply: active\n");

	sigaction(SIGINT, &term_action, NULL);
	sigaction(SIGCHLD, &term_action, NULL);

	if (cmd) {
		int err = 0;

		if (write(inftrig, &err, sizeof(err)) != sizeof(err)) {
			fprintf(stderr, "ply: unable to start command\n");
			ret.err = 1;
			ret.val = -EIO;
			goto stop;
		}
	}

	ret = ply_loop(ply);
	if (ret.err && (ret.val == EINTR) && term_sig)
		ret.err = 0;
stop:
	_d("ply: deactivating\n");
	ply_stop(ply);

	ply_maps_print(ply);

unload:
	ply_unload(ply);

err:
	ply_free(ply);

	if (ret.err) {
		if (ret.val)
			printf("ERR:%d\n", ret.val);

		return 1;
	}

	return ret.exit ? ret.val : 0;
}
